### PR TITLE
✨ Add stripe cards admin page

### PR DESCRIPTION
**Pull request by malted**
Narrator: | According to all known laws of aviation, there is no way that a bee should be able to fly. Its wings are too small to get its fat little body off the ground. The bee, of course, flies anyway because bees don't care what humans think is impossible.
-- | --

Narrator:	According to all known laws of aviation, there is no way that a bee should be able to fly. Its wings are too small to get its fat little body off the ground. The bee, of course, flies anyway because bees don't care what humans think is impossible.
 	Cut to Barry's room, where he's picking out what to wear.
[Barry](https://beemovie.fandom.com/wiki/Barry)	Yellow, black. Yellow, black. Yellow, black. Yellow, black. Ooh, black and yellow! Yeah, let's shake it up a little.
 	Barry uses honey from a dispenser to style his hair, rinse his mouth, and then applies it to his armpits.
Mom (Janet Benson)	(calling from downstairs:) Barry! Breakfast is ready!
Barry:	Coming! (phone rings) Oh, hang on a second. (adjusts his antennas into a headset) Hello?
[Adam Flayman](https://beemovie.fandom.com/wiki/Adam_Flayman)	(on the phone) Barry?
Barry:	Adam?
Adam:	Can you believe this is happening?
Barry:	I can't believe it. I'll pick you up. (hangs up, sharpens his stinger) Lookin' sharp. (flies downstairs)
Mom:	Barry, why don't you use the stairs? Your father paid good money for those.
Barry:	Sorry. I'm excited.
[Dad](https://beemovie.fandom.com/wiki/Dad) (Martin Benson):	Here's the graduate. We're very proud of you, son. And a perfect report card, all B's.
Mom:	Very proud. (touches Barry's hair)
Barry:	Ma! I got a thing going here.
Mom:	Ah, you got some lint on your fuzz.
Barry:	Ow! That's me!
Dad:	Wave to us! We'll be in row 118,000.
Barry:	Bye! (flies off)
Mom:	Barry, I told you, stop flying in the house!
 	(Barry drives his car to pick up his classmate. Adam's outside his house, reading the Hive Today newspaper. The front page headline reads "FRISBEE HITS HIVE ! Internet Down. Bee: 'I heard sound, then Wham-o!'")
Barry:	Hey, Adam.
Adam:	Hey, Barry. Is that fuzz gel?
Barry:	A little. It's a special day, finally graduating.
Adam:	Never thought I'd make it.
Barry:	Yeah, three days of grade school, three days of high school.
Adam:	Those were so awkward.
Barry:	Three days of college. I'm glad I took off one day in the middle and just hitchhiked around the hive.
Adam:	You did come back different.
 	(a bee calls out as they drive past)
Bee:	Hi, Barry.
Barry:	Hey Artie, growing a mustache? Looks good.
Adam:	Hey, did you hear about Frankie?
Barry:	Yeah.
Adam:	You goin' to his funeral?
Barry:	No, I'm not goin' to his funeral. Everybody knows you sting someone, you die. You don't waste it on a squirrel. He was such a hothead.
Adam:	Yeah, I guess he could have just gotten out of the way.
 	(They make various noises as the car goes up and down some hills and does a loop on the road.)
A & B	Woah! Oooooooh!
Adam:	I love this incorporating an amusement park right into our regular day.
Barry:	I guess that's why they say we don't need vacations.
 	(They arrive, fly in and take their seats.)
Barry:	Boy, quite a bit of [pomp... under the circumstances](http://en.wikipedia.org/wiki/Pomp_and_Circumstance_Marches).
Barry:	Well, Adam, today we are men.
Adam:	We are!
Barry:	Bee-men.
Adam:	Amen!
A & B:	Hallelujah! (bumping each other) Aaaaaaaaaaaah!
Announcer:	Students, faculty, distinguished bees, please welcome Dean Buzzwell.
 	Dean Buzzwell walks onto the stage and taps the microphone.
Buzzwell:	Welcome, New Hive City graduating class of... (presses a button to change the timer on the podium from 9:00 to 9:15) ...9:15. And that concludes our graduation ceremonies.
 	(Students cheer, throw their caps into the air as helmets are placed on their heads.)
Buzzwell:	And begins your career at [Honex Industries](https://beemovie.fandom.com/wiki/Honex_Industries)!
Barry:	Are we gonna pick our jobs today?
Adam:	I heard it's just orientation.
Barry:	Huh. Woah. Heads up! Here we go.
 	(The stands for Winger University the students are sitting in begin converting into tram seating.)
Female announcer:	Keep your hands and antennas inside the tram at all times. (flies down to go in the tram as it starts moving and repeats it in Spanish:) Mantenga sus manos y antenas dentro del tranvía en todo momento.
Barry:	Wonder what it's going to be like?
Adam:	A little scary. (he and Barry mimic shivering and making scared noises)
[Trudy](https://beemovie.fandom.com/wiki/Trudy), the Honex tour guide:	Welcome to [Honex](https://beemovie.fandom.com/wiki/Honex), a division of Honesco and a part of the Hexagon Group.
Barry:	This is it!
everyone:	Wow. (tram moves into the factory floor)
Barry:	Wow.
Trudy:	We know that you, as a bee, have worked your whole life to get to the point where you can work for your whole life. Honey begins when our valiant Pollen Jocks bring the nectar to the hive. Our top-secret formula is automatically color-corrected, scent-adjusted and bubble-contoured into this soothing sweet syrup with its distinctive golden glow you know as...
everyone:	Honey!
 	(Tour guide grabs a beaker of honey as they drive by and tosses it to the group, which bounce it around towards the back.}
Adam:	That girl was hot.
Barry:	She's my cousin!
Adam:	She is?
Barry:	Yes, we're all cousins.
Adam:	Right. You're right.
Trudy:	At Honex, we also constantly strive to improve every aspect of bee existence. These bees are stress-testing a new helmet technology.
 	(Behind a display window, a bee puts on a helmet, then runs back and forth as levers holding a rolled-up magazine, flyswatter and a shoe move down to try hit him. He is hit by the magazine, dodges the flyswatter, but then hit by the boot and again by all three, followed by being sprayed with aerosol from two cans. He signals he's okay, but is flattened by the flyswatter, magazine and shoe converging to strike him together. He signals he's still okay by poking his arm up through a hole in the flyswatter and giving another thumbs up. The tram riders applaud.)
Adam:	Ooh. What do you think he makes?
Barry:	Not enough.
Trudy:	And here we have our latest advancement, the [Krelman](https://beemovie.fandom.com/wiki/Krelman).
Barry:	Wow, what does that do?
Trudy:	Catches that little strand of honey that hangs after you pour it. Saves us millions.
 	(A Krelman worker waves and Adam waves back.)
Adam:	Uh, uh, can anyone work on the Krelman?
Trudy:	Of course. Most bee jobs are small ones. But bees know that every small job, if it's done well, means a lot. But choose carefully because you'll stay in the job that you pick for the rest of your life.
Barry:	The same job the rest of your life? I didn't know that.
Adam:	What's the difference?
Barry:	Huh?
Trudy:	And you'll be happy to know that bees, as a species, haven't had one day off in 27 million years. Wooh!
Barry:	So you'll just work us to death?
Trudy:	We'll sure try.
 	(Everyone laughs while Barry looks uncomfortable. The tram converts into a boat that flows down a log-flume style ramp with honey in it, then converts back to a wheeled tram at the end.)
 	(With the tour over, Adam and Barry head home. Adam jumps with excitement.)
Adam:	Wow! That blew my mind!
Barry:	"What's the difference?" Adam, how could you say that? One job forever? That's an insane choice to have to make.
Adam:	Well, I'm relieved. Now we only have to make one decision in life.
Barry:	But, Adam, how could they never have told us that?
Adam:	Barry, why would you question anything? We're bees. We're the most perfectly functioning society on Earth.
 	(a filling station attendant yells at a bee for putting the honey nozzle into his own mouth.)
Barry:	Yeah, but Adam, did you ever think that maybe things work a little too well around here?
Adam:	Like what? Give me one example.
 	(both stop in the middle of an intersection. the traffic adjusts to drive around them.)
Barry:	I don't know. But you know what I'm talking about.
Announcer over speaker:	Please clear the gate. Royal Nectar Force on approach. Royal Nectar Force on approach.
Barry:	Wait a second. Check it out. Hey, those are [Pollen Jocks](https://beemovie.fandom.com/wiki/Pollen_Jock)!
Adam:	Wow.
 	(Pollen Jocks fly into the hive and land.)
Adam:	I've never seen them this close.
Barry:	They know what it's like to go outside the hive.
Adam:	Yeah, but some of them don't come back.
 	(two lady bees wave at the jocks and call out:)
Girl Bees	Hey, Jocks!
Hi, Jocks!
 	(the pollen is removed from the jocks and collected into storage capsules marked "Nectar", then trucked away. A General flies over to welcome them.)
[General](https://beemovie.fandom.com/wiki/General):	You guys did great! You're monsters! You're sky freaks! I love it! I love it!
Barry:	I wonder where those guys have just been.
Adam:	I don't know.
Barry:	Their day's not planned. Outside the hive, flying who knows where, doing who knows what.
Adam:	You can't just decide one day to be a Pollen Jock. You have to be bred for that.
Barry:	Right. (pollen begins drifting down around them) Look at that. That's more pollen than you and I will ever see in a lifetime.
Adam:	It's just a status symbol. I think bees make too big a deal out of it.
Barry:	Perhaps. Unless you're wearing it and the ladies see you wearing it.
 	(the same two lady bees giggle at being talked about by Barry)
Adam:	Those ladies? Aren't they our cousins too?
Barry:	Distant. Distant.
 	Three pollen jocks observe them.
Jackson:	Look at these two.
Splitz:	Couple of Hive Harrys.
Jackson:	Let's have some fun with them.
Lady 1:	It must be so dangerous being a Pollen Jock.
Barry:	Oh, yeah. One time a bear had me pinned up against a mushroom! He had one paw on my throat, and with the other, he was slapping me back and forth across the face!
Lady 2:	Oh, my!
Barry:	I never thought I'd knock him out.
Lady 1:	(to Adam) And what were you doing during all of this?
Adam:	Obviously, I was trying to alert the authorities.
Barry:	I can autograph that if you want.
Jackson:	A little gusty out there today, wasn't it, comrades?
Barry:	Yeah. Gusty.
Buzz:	Yeah, we're gonna hit a sunflower patch about six miles from here tomorrow.
Barry:	Six miles, huh?
Adam:	Barry!
Buzz:	It's a puddle jump for us, but, uh, maybe you're not up for it.
Barry:	Maybe I am.
Adam:	(quietly:) You are not.
Buzz:	We're going 0900 at J-Gate.
Adam:	Woah!
Buzz:	What do you think, buzzy-boy? Are you bee enough?
Barry:	I might be. It all depends on what 0900 means.
 	(later, back at home that night, Barry is on the balcony, looking out at the hive)
Dad:	Hey, Honex!
Barry:	Oh! Dad, you surprised me.
Dad	(laughing) Have you decided what you're interested in, son?
Barry:	Well, there's a lot of choices.
Dad:	But you only get one. (laughs again)
Barry:	Dad, do you ever get bored doing the same job every day?
Dad:	Son, let me tell you something about stirring. You grab that stick, and you just move it around, and you stir it around. You get yourself into a rhythm. It's a beautiful thing.
Barry:	You know, Dad, the more I think about it, maybe the honey field just isn't right for me.
Dad:	And you were thinking of what, making balloon animals? That's a bad job for a guy with a stinger.
Barry:	Well, no...

### CODE REVIEW

File: app/views/admin/stripe_cards.html.erb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    <tr>
+      <th class="w-16">Last 4</th>
+      <th class="w-32">Created</th>
+      <th>Name</th>
+      <th>Type</th>
+    </tr>
+  </thead>
+  <tbody>
+    <% @cards.each do |card| %>
+      <tr style="<%= "background: #ff8c38;" if user.admin_at.present? %>">
```
Comment (garyhtou) [inline app/views/admin/stripe_cards.html.erb]:
```suggestion
      <tr>
```

I think this code came from the admin users page and was intended to highlight a row if the user is an admin. Since this page is for Stripe Cards and `user` isn't defined here, let's remove this logic

File: app/models/event.rb [hunk 2]
```diff
@@ -49,7 +49,7 @@
 #  club_airtable_id                :text
 #  emburse_department_id           :string
 #  increase_account_id             :string           not null
-#  partner_id                      :bigint           not null
+#  partner_id                      :bigint
```
Comment (garyhtou) [inline app/models/event.rb]: Let's remove this unrelated change

File: app/views/admin/stripe_cards.html.erb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+<h1>STRIPE CARDS</h1>
+
+<table>
+  <thead>
+    <tr>
+      <th class="w-16">Last 4</th>
+      <th class="w-32">Created</th>
+      <th>Name</th>
+      <th>Type</th>
```
Comment (garyhtou) [inline app/views/admin/stripe_cards.html.erb]: non-blocking feedback:

It would also be helpful to include the following:
- Cardholder's name: `card.cardholder.user.name`
- Link to the card's page: `stripe_card_url(card)`

**Review by garyhtou — COMMENTED** [review]
_According to all known laws of aviation, this is a great PR!_

I left a couple of comments, some of them are non-block and it's totally up to you whether you want to implement them.

---

Also, can we add this page to the already super duper long admin navigation?
image

https://github.com/hackclub/bank/blob/f0df6c2ec699db7443bb7cdb2d2c6c96edbb0229/app/views/layouts/admin.html.erb#L31-L80

File: app/controllers/admin_controller.rb [hunk 4]
```diff
@@ -312,6 +312,14 @@ def users
     render layout: "admin"
   end
 
+  def stripe_cards
+
+    @cards = StripeCard.all
```
Comment (northeastprince) [inline app/controllers/admin_controller.rb]: No pagination?

Comment (malted) [reply app/controllers/admin_controller.rb]: Added!

File: app/views/admin/stripe_cards.html.erb (hunk 1 above)
Comment (malted) [reply app/views/admin/stripe_cards.html.erb]: Done!

File: app/views/admin/stripe_cards.html.erb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+      <th>Card name</th>
+      <th>Type</th>
+      <th>URL</th>
+    </tr>
+  </thead>
+  <tbody>
+    <% @cards.each do |card| %>
+      <tr>
+        <td><%= card.cardholder.user.name %>
+        <td><%= card.formatted_card_number %></td>
```
Comment (cjdenio) [inline app/views/admin/stripe_cards.html.erb]: Maybe use `card.hidden_card_number_with_last_four` instead?
- `formatted_card_number` makes a Stripe API request so it's slow
- for security, it's probably best to not show full card numbers (even to admins)

Comment (malted) [reply app/views/admin/stripe_cards.html.erb]: Ah. Ok!
I had that originally anyway.

Why is it not ok to show it to admins? They can access them through the stripe dash anyway right?

File: app/controllers/admin_controller.rb [hunk 6]
```diff
@@ -312,6 +312,16 @@ def users
     render layout: "admin"
   end
 
+  def stripe_cards
+    @page = params[:page] || 1
+    @per = params[:per] || 20
+
+    @cards = StripeCard.all.page(@page).per(@per).order("created_at desc")
```
Comment (cjdenio) [inline app/controllers/admin_controller.rb]:
```suggestion
    @cards = StripeCard.includes(:stripe_cardholder).page(@page).per(@per).order("created_at desc")
```
Slightly improves performance: see https://guides.rubyonrails.org/active_record_querying.html#eager-loading-associations

File: app/views/admin/stripe_cards.html.erb (hunk 5 above)
Comment (cjdenio) [reply app/views/admin/stripe_cards.html.erb]: I think admins should _have_ to go into the Stripe dashboard to view card numbers. There are a lot of Bank admins and it opens us up to liability issues if anyone associated with the Bank team has easy access to users' card numbers

Comment (malted) [reply app/views/admin/stripe_cards.html.erb]: Got it, ty!

Comment (malted) [reply app/views/admin/stripe_cards.html.erb]: Displaying last 4 now

**Review by cjdenio — APPROVED** [review]
This looks great to me, thanks! We should also consider adding search at some point
